configure.ac

```diff
@@ -28,9 +28,49 @@ AC_CONFIG_MACRO_DIR([m4/autoconf])
 
 m4_ifdef([AM_SILENT_RULES], [AM_SILENT_RULES([yes])])
 
-# To have empty CFLAGS instead of undefined and '-g -O2' by default
 CFLAGS=$CFLAGS
-CFLAGS="-Os -Wall -Wextra -Werror $CFLAGS"
+# To have empty CFLAGS instead of undefined and '-g -O2' by default
+
+### HOST FOR CROSS-COMPILING
+target_os=`uname -s`
+
+AC_MSG_CHECKING([target os])
+
+case "$target_os" in
+	MINGW* | cygwin* | msys* | mingw* | cegcc*)
+		CFLAGS="-g -g3 -Os -Wall -Wextra -std=c99 -Wformat $CFLAGS"
+		LDFLAGS="$LDFLAGS -Wl,--no-undefined"
+		target_os=mingw
+    ;;
+	ANDROID*)
+		target_os=android
+		CFLAGS="-Os -fPIE  -Wall -Wextra  $CFLAGS"
+		LDFLAGS="$LDFLAGS -pie"
+    ;;
+	    *)
+		target_os=linux
+		CFLAGS="-Os -Wall -Wextra -Werror $CFLAGS"
+AC_MSG_RESULT([$target_os])
+    ;;
+esac
+AC_MSG_RESULT([$target_os])
+
+AC_MSG_CHECKING([checking for obstack_header])
+AC_CHECK_HEADERS([obstack.h],
+        [found_obstack_headers=yes; break;])
+AS_IF([test "x$found_obstack_headers=yes" != "xyes"],
+[AC_MSG_RESULT([Unable to find the obstack headers])])
+AM_CONDITIONAL(ADD_OBSTACK, [test x$found_obstack_headers != xyes])
+
+AC_MSG_CHECKING(whether program_invocation_short_name is defined)
+AC_TRY_COMPILE([#include <argp.h>],
+               [program_invocation_short_name = "test";],
+               AC_DEFINE(HAVE_PROGRAM_INVOCATION_SHORT_NAME, 1,
+                [Define if program_invocation_short_name is defined])
+                AC_MSG_RESULT(yes),
+               AC_MSG_RESULT(no))
+
+
```

Comment:
The rest is self explanatory:
1. check for obstack.h header
a. set autotools variable 
b. write macro to config.h if not existing

2 test program_invocation function
a. call arg1 in main otherwise or leave as string in libraries